x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -208,42 +232,78 @@ private void searchEventGroupedByStackTrace(
     ) {
         GetStackTracesResponseBuilder responseBuilder = new GetStackTracesResponseBuilder();
         responseBuilder.setSamplingRate(eventsIndex.getSampleRate());
+        responseBuilder.setRequestedDuration(request.getRequestedDuration());
         client.prepareSearch(eventsIndex.getName())
             .setTrackTotalHits(false)
             .setSize(0)
             .setQuery(request.getQuery())
             .addAggregation(new MinAggregationBuilder("min_time").field("@timestamp"))
             .addAggregation(new MaxAggregationBuilder("max_time").field("@timestamp"))
             .addAggregation(
+                // We have nested aggregations, which in theory might blow up to 150.000 * 20.000 items
+                // reported. But we know that the total number items is limited by our down-sampling to
+                // a maximum of 100k (plus a bit due to rounding errors).
                 new TermsAggregationBuilder("group_by")
-                    // 'size' should be max 100k, but might be slightly more. Better be on the safe side.
-                    .size(150_000)
-                    .field("Stacktrace.id")
+                    // 'size' specifies the max number of host ID we support per request.
+                    .size(maxSupportedUniqueHosts)
+                    .field("host.id")
                     // 'execution_hint: map' skips the slow building of ordinals that we don't need.
                     // Especially with high cardinality fields, this makes aggregations really slow.
                     .executionHint("map")
-                    .subAggregation(new SumAggregationBuilder("count").field("Stacktrace.count"))
+                    .subAggregation(
+                        new TermsAggregationBuilder("group_by")
+                            // 'size' should be max 100k, but might be slightly more. Better be on the safe side.
+                            .size(150_000)
+                            .field("Stacktrace.id")
+                            // 'execution_hint: map' skips the slow building of ordinals that we don't need.
+                            // Especially with high cardinality fields, this makes aggregations really slow.
+                            .executionHint("map")
+                            .subAggregation(new SumAggregationBuilder("count").field("Stacktrace.count"))
+                    )
             )
             .addAggregation(new SumAggregationBuilder("total_count").field("Stacktrace.count"))
             .execute(handleEventsGroupedByStackTrace(client, responseBuilder, submitListener, searchResponse -> {
-                StringTerms stacktraces = searchResponse.getAggregations().get("group_by");
-                Sum totalCountAgg = searchResponse.getAggregations().get("total_count");
-                long totalCount = Math.round(totalCountAgg.value());
+                long totalCount = getLongFromResponse(searchResponse, "total_count");
```

Comment:
IMHO that method does not convey the intent (which long? which part of the response?) clearly enough. How about `long getAggValueAsLong(String aggregationName)`?